src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReadsDownsamplingIterator.java

```diff
@@ -0,0 +1,91 @@
+package org.broadinstitute.hellbender.utils.downsampling;
+
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Iterator wrapper around our generic {@link ReadsDownsampler)} interface. Converts the push-style
+ * downsampler interface to a pull model.
+ */
+public final class ReadsDownsamplingIterator implements Iterator<GATKRead>, Iterable<GATKRead> {
+
+    private final Iterator<GATKRead> nestedReadIterator;
+    private final ReadsDownsampler downsampler;
+    private Collection<GATKRead> downsampledReadsCache;
```

Comment:
why is `downsampledReadsCache` an instance field? It's only used within 1 method